### PR TITLE
[FLINK-21092][FLINK-21093][FLINK-21094][FLINK-21096][table-planner-blink] Support ExecNode plan serialization/deserialization for `INSERT INTO MySink SELECT * FROM MyTable` 

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*This pr aims to support getting/executing exec node plan for SQL: `INSERT INTO MySink SELECT * FROM MyTable` *

## Brief change log
  - *Introduce getJsonPlan, explainJsonPlan and executeJsonPlan in TableEnvironmentInternal*
  - *Support StreamExecTableSource json serialization/deserialization*
  - *Support StreamExecSink json serialization/deserialization*
  - *Introduce ExecNodeGraphJsonPlanGenerator to serialize ExecNodeGraph to json plan and deserialize json plan to ExecNodeGraph*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests to verify the serialization/deserialization for different classes, such as LogicalTypeSerdeTest, ExecEdgeSerdeTest*
  - *Extended integration test to verify getJsonPlan, explainJsonPlan and executeJsonPlan on `INSERT INTO MySink SELECT * FROM MyTable` statement *

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by godfreyhe** [comment]
cc @wenlong88

**Comment by godfreyhe** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/CatalogTableSpecBase.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** {@link CatalogTableSpecBase} describes how to serialize/deserialize a catalog table. */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CatalogTableSpecBase {
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/CatalogTableSpecBase.java:51]: move spec class to utils or create a new packages for specs?

**Review by wenlong88 — COMMENTED** [review]
Thanks for the contribution, I have just reviewed and left some comments on the first two commit.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/ExecEdge.java [hunk 2]
```diff
@@ -21,39 +21,97 @@
...
 
     public static final ExecEdge DEFAULT = ExecEdge.builder().build();
 
+    public static final String FIELD_NAME_REQUIRED_SHUFFLE = "requiredShuffle";
+    public static final String FIELD_NAME_DAM_BEHAVIOR = "damBehavior";
+    public static final String FIELD_NAME_PRIORITY = "priority";
+
+    @JsonProperty(FIELD_NAME_REQUIRED_SHUFFLE)
+    @JsonSerialize(using = RequiredShuffleJsonSerializer.class)
+    @JsonDeserialize(using = RequiredShuffleJsonDeserializer.class)
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/ExecEdge.java:58]: json annotation is enough for ExecEdge now, but I do some refactoring in #14757, then it need a specific serializer/deserializer. I can revert them and introduce them after #14757 is finished.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/LogicalTypeJsonSerializer.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            jsonGenerator.writeEndObject();
+        } else if (logicalType instanceof TypeInformationRawType) {
+            // TypeInformationRawType does not support `asSerializableString`
+            TypeInformationRawType<?> rawType = (TypeInformationRawType<?>) logicalType;
+            jsonGenerator.writeStartObject();
+            jsonGenerator.writeBooleanField(FIELD_NAME_NULLABLE, rawType.isNullable());
+            jsonGenerator.writeStringField(
+                    FIELD_NAME_TYPE_INFO,
+                    EncodingUtils.escapeSingleQuotes(
+                            TypeStringUtils.writeTypeInfo(rawType.getTypeInformation())));
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/LogicalTypeJsonSerializer.java:70]: refer to `LegacyTypeInformationType`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/DynamicTableSourceSpec.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import java.io.IOException;
+
+/**
+ * {@link DynamicTableSourceSpec} describes how to serialize/deserialize dynamic table source table
+ * and create {@link DynamicTableSource} from the deserialization result.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class DynamicTableSourceSpec extends CatalogTableSpecBase {
+
+    @JsonIgnore private DynamicTableSource tableSource;
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/DynamicTableSourceSpec.java:50]: good catch! we should provide the a new interface to serialize push-down part in the table source, and when doing deserialization, we will build the table source instance first, and then apply the push-down to the new table source.

File: flink-table/flink-table-planner-blink/pom.xml [hunk 5]
```diff
@@ -418,6 +419,11 @@ under the License.
 									<shadedPattern>org.apache.flink.table.shaded.org.codehaus</shadedPattern>
 								</relocation>-->
 
+								<relocation>
+									<pattern>org.reflections</pattern>
+									<shadedPattern>org.apache.flink.calcite.shaded.org.reflections</shadedPattern>
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/pom.xml:424]: typo, I will fix it.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/ExecNodeBase.java [hunk 6]
```diff
@@ -36,46 +46,97 @@
...
+        return id;
+    }
+
+    /** The name used to identify each sub-class in the json plan. */
+    @JsonProperty(value = FIELD_NAME_CLASS, access = JsonProperty.Access.READ_ONLY)
+    public final String getClassName() {
+        return getClass().getCanonicalName();
+    }
+
+    @JsonProperty(value = FIELD_NAME_DESCRIPTION)
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/ExecNodeBase.java:109]: there are no fields for `getClassName` and `getInputNodes`(`inputNodes` will be removed later), putting `JsonProperty` all on `get` method will more unified style.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/CatalogTableSpecBase.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+    @JsonProperty(value = FIELD_NAME_CATALOG_TABLE, required = true)
+    @JsonSerialize(using = CatalogTableJsonSerializer.class)
+    @JsonDeserialize(using = CatalogTableJsonDeserializer.class)
+    protected final CatalogTable catalogTable;
+
+    @JsonProperty(value = FIELD_NAME_CONFIGURATION, required = true)
+    @JsonSerialize(using = ReadableConfigJsonSerializer.class)
+    @JsonDeserialize(using = ReadableConfigJsonDeserializer.class)
+    protected final ReadableConfig configuration;
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/CatalogTableSpecBase.java:73]: yes, we can persist configuration once on the top level of the json plan

**Review by wenlong88 — COMMENTED** [review]
thanks for the contribution, the overall change LGTM, just a minor comment left.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/StatementSetImpl.java [hunk 8]
```diff
@@ -100,4 +101,27 @@ public TableResult execute() {
...
+     *
+     * @return the string json representation of an optimized ExecNode plan for the statements and
+     *     Tables.
+     */
+    @Experimental
+    public String getJsonPlan() {
+        try {
+            return tableEnvironment.getJsonPlan(operations);
+        } finally {
+            operations.clear();
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/StatementSetImpl.java:124]: make sense

**Review by wenlong88 — APPROVED** [review]
LGTM

**Pull request closed without merging**